strings/src/main/java/org/odk/collect/strings/LocalizedApplication.kt

```diff
@@ -0,0 +1,31 @@
+package org.odk.collect.strings
+
+import android.content.Context
+import android.content.res.Configuration
+import java.util.Locale
+
+interface LocalizedApplication {
+
+    val locale: Locale
+}
+
+fun Context.getLocalizedString(stringId: Int, vararg formatArgs: Any): String {
+    return when (applicationContext) {
+        is LocalizedApplication -> {
```

Comment:
You won't be surprised to hear me say that as cool as extensions are, they make me nervous. It just feels unexpected to have modifications to `Collect` happening way over here. In this specific case, can the actual `getLocalizedString` definition go in `Collect` and the extension only handle the case where it doesn't exist?

Reply:
This is extending `Context`, not `Collect` (although I just typed that wrong way round twice 😂). `strings` is a dependency of `collect_app` so this is just adding a helper that any `Context` object can use rather than modifying a concrete type we own.

 I don't know if you spotted it, but in Java you don't end up with `Content#getLocalizedString(Int, Any)` you get  `getLocalizedString(Context, Int, Any)` (so a static with the first argument as the "receiver"). Extensions in Kotlin are just sugar and don't modify the interface of their receivers (so they can't access private state for example). Does that all still make you nervous? My Kotlin is all coming back to me pretty fast so definitely happy to tone it down! We could just make this a static in both Kotlin and Java.

Reply:
> This is extending Context, not Collect

`Collect` is the only implementation that would go through the first branch because it's the only `Context` that implements `LocalizedApplication`, right?

> you get getLocalizedString(Context, Int, Any)

And you can pass any `Context` in and it will "just work" but actually you'll only get the correct strings if you pass in a `LocalizedApplication`?

Reply:
Now I'm super confused. Won't the usage made in `TranslationHandler` not work because generally the context passed in will be a fragment or activity which will go through to the else branch?

Reply:
> Collect is the only implementation that would go through the first branch because it's the only Context that implements LocalizedApplication, right?

Right. I originally looked at moving the `TranslationHandler` into `strings` but it ends up calling `LocaleHelper` which is currently pretty wrapped up in our preferences. So my thought was to separate out the logic of loading localized resources from however the preference of the locale was stored/defined.

> And you can pass any Context in and it will "just work" but actually you'll only get the correct strings if you pass in a LocalizedApplication?

Yeah. In my head the story is that `strings` supports lots of languages but it doesn't know if its dependents actually provide resource localization at an application level (rather than a system level). Also, it means we don't have to have our test Application objects care about provide a locale or implementing `LocalizedApplication` I can totally accept that I'm trying to over architect myself out of the hax that makes all this work (the messing with context to fool Android into loading different resources).

Reply:
> Now I'm super confused. Won't the usage made in TranslationHandler not work because generally the context passed in will be a fragment or activity which will go through to the else branch?

Sorry just saw this. It's branching on `applicationContext` (from `Context`) so it should get the application object in those cases.

Reply:
I had totally missed that. I'm not very familiar with Kotlin so may not be best positioned for some of these reviews. I thought `applicationContext` was just the name you were giving to the context being extended.

Reply:
Ah it also looks like we're not being helped by Github's syntax highlighting here. Extension functions have `this` as an implicit "context" and then Kotlin is using its property access sugar to "sweeten" up the call to `getApplicationContext()`. In IntelliJ/Android Studio it highlights as a field but Github isn't giving us that ☹️.